testing/base.py

```diff
@@ -95,29 +105,84 @@ def __new__(mcs, name, bases, attrs):
         if has_tests:
+            ############################################################################
+            # Make the sphinx test application available as `self.app`.                #
+            ############################################################################
             def _set_app(self, app):
                 # before the test
                 self.app = app
                 yield  # the test runs
-                # cleanup after the test: exhale and sphinx output dirs and breathe project dirs
-                dirs = [app.config.exhale_args['containmentFolder'], os.path.dirname(app.outdir)]
-                for d in app.config.breathe_projects.values():
-                    # get only first directory of breathe project directory relative to srcdir
-                    dirs.append(os.path.relpath(os.path.join(app.srcdir, d), app.srcdir).split(os.sep)[0])
-                for d in dirs:
-                    if not os.path.isabs(d):
-                        # d is a relative dir, make it absolute
-                        d = os.path.join(app.srcdir, d)
-                    shutil.rmtree(d, ignore_errors=True)
+                # This cleanup happens between each test case, do not delete docs/
+                # until all tests for this class are done!
+                containmentFolder = self.getAbsAgainstSrcdir("containmentFolder")
+                if os.path.isdir(containmentFolder):
+                    shutil.rmtree(containmentFolder)
+                # Delete the doctrees as well as e.g. _build/html, app.outdir is going
+                # to be docs/_build/{builder_name}
+                _build = os.path.abspath(os.path.dirname(app.outdir))
+                if os.path.isdir(_build):
+                    shutil.rmtree(_build)
+                # Make sure doxygen output is deleted between runs
+                doxy_xml_dir = os.path.abspath(os.path.join(
+                    self.app.srcdir, app.config.breathe_projects[test_project]
+                ))
+                doxy_dir = os.path.dirname(doxy_xml_dir)
+                if os.path.isdir(doxy_dir):
+                    shutil.rmtree(doxy_dir)
+
             attrs['_set_app'] = pytest.fixture(autouse=True)(_set_app)
 
         # applying the default configuration override, which is overridden using the @confoverride decorator
         # at class or method level
-        return default_confoverrides(
+        cls = default_confoverrides(
             super(ExhaleTestCaseMetaclass, mcs).__new__(mcs, name, bases, attrs),
             make_default_config(attrs['test_project'])
         )
 
+        if has_tests:
+            ############################################################################
+            # Automatically create docs/{conf.py,index.rst} for the test project.      #
+            ############################################################################
+            def _with_docs_dir(self):
```

Comment:
Previously before `return default_confoverrides` I was setting a class-level `cls.config` variable (hence the question earlier).  Locally, with either `scope="function"` or `scope="class"` the `testing/projects/{test_project}/docs/{conf.py,index.rst}` are currently being created, but not on Travis for some reason.  Locally they are not deleted though.

I had done `cls = default_confoverrides` and bound `_with_docs_dir` later because without it `self.app.config` is not available.  So it seems there is a more glaring problem that I will investigate further.

The idea is in the future I will want to be able to actually save all of the generated reStructuredText and build out the testing website.  I'm probably going to do it on a `gh-pages` branch (far far off in the future) because it will be very useful to actually set `GENERATE_HTML = YES` to have the ground truth (Doxygen) included.

Reply:
The way to change the sphinx root directory for the ``sphinx.testing.fixtures.app`` is to use ``pytest.mark.parametrize``. Some digging will be needed to know where and with which arguments to call it.